core/include/gpio.h

```diff
@@ -23,16 +26,42 @@ enum gpio_interrupt {
 
 struct gpio_chip {
 	const struct gpio_ops *ops;
+	unsigned int gpio_base;
+	unsigned int ngpios;
+	vaddr_t base;
+
+	SLIST_ENTRY(gpio_chip) link;
 };
 
 struct gpio_ops {
-	enum gpio_dir (*get_direction)(unsigned int gpio_pin);
-	void (*set_direction)(unsigned int gpio_pin, enum gpio_dir direction);
-	enum gpio_level (*get_value)(unsigned int gpio_pin);
-	void (*set_value)(unsigned int gpio_pin, enum gpio_level value);
-	enum gpio_interrupt (*get_interrupt)(unsigned int gpio_pin);
-	void (*set_interrupt)(unsigned int gpio_pin,
-		enum gpio_interrupt ena_dis);
+	enum gpio_dir (*get_direction)(struct gpio_chip *gc, unsigned int pin);
+	void (*set_direction)(struct gpio_chip *gc, unsigned int pin,
+			      enum gpio_dir direction);
+	enum gpio_level (*get_value)(struct gpio_chip *gc, unsigned int pin);
+	void (*set_value)(struct gpio_chip *gc, unsigned int pin,
+			  enum gpio_level value);
+	enum gpio_interrupt (*get_interrupt)(struct gpio_chip *gc,
+					     unsigned int pin);
+	void (*set_interrupt)(struct gpio_chip *gc, unsigned int pin,
+			      enum gpio_interrupt ena_dis);
+};
```

Comment:
This will break `pl061_gpio.c`!

Reply:
You might not have these platforms to try on, but please confirm that at least the building of this for hikey and poplar platforms are not broken.

Reply:
Thanks.  I did see the current users those will get affected by this change.
the following two GPIO drivers.
1-core/drivers/stm32_gpio.c
2-core/drivers/pl061_gpio.c
3-And their users ie currently only one spi driver core/drivers/pl022_spi.c  << this change is already in this series. And I am testing this path by a bcm_test_pta -> pl022_spi -> gpio -> bcm_gpio
3-Users of spi driver ..  I think only plat-hikey spi_test.  

Once the base is finalized. It will be one time work to move the existing GPIO drivers on to this.
I will try to add a commit to pass compilation of the affected platforms.

Reply:
You're welcome. Not sure about stm32 but definitely hikey and poplar are affected since they include `pl061_gpio.c`. For reviewing you can (should) add a commit to pass compilation of all affected platforms, but before merge, this additional commit should be squashed into this commit, else we'd have a broken commit in history.

Reply:
Alright. Let me see the affected drivers in detail and get back.

Reply:
Are you able to show the bcm_test_pta code to see how this framework is actually used?

Reply:
sure . the PTA  are not in clean shape.  I will upload a reference pta soon.  Also the current patch set have spi driver pl022_spi.c  using the gpio apis.

Reply:
Thanks! If you have it hosted on github somewhere, please provide the url. It doesn't matter the shape atm. Just want to compare it against the current usage to see if how much easier or more difficult it is to use the proposed framework. pl022_spi.c is using the gpio apis, but not the whole setup end-to-end/top-to-bottom.

Reply:
nothing more in bcm test pta
```
#include <gpio.h>
gpio_desc  *gd;
gd = gpio_request(pin);   or  gpio_request(pin, data);  /* I am removing the owner field as per review.  the usage will come when we add interrupt handling to it */
rest of the apis as ..
  gpio_xxx(gd, <xxx>); 
gpio_release(pin);
```
bcm gpio test pta
https://github.com/sandeepbrcm/optee_os/blob/8e3f191353d371ba7cc4228e0e242c34fd691834/core/arch/arm/pta/bcm_pta/src/pta_bcm_gpio.c#L180-L182

Reply:
thanks will take a look